src/multicalendarstorage.cpp

```diff
@@ -0,0 +1,532 @@
+}
+
+//@cond PRIVATE
+class MultiCalendarStorage::Private: public AlarmHandler
+{
+public:
+    Private(MultiCalendarStorage *storage, const QTimeZone &timezone)
+        : mStorage(storage)
+        , mTimezone(timezone)
+    {}
+
+    MultiCalendarStorage *mStorage;
+    QList<MultiCalendarStorage::Observer *> mObservers;
+    QHash<QString, CalendarHandler> mCalendars;
+    QString mDefaultNotebookUid;
+    QTimeZone mTimezone;
+    QList<Range> mRanges;
+    bool mIsRecurrenceLoaded = false;
+
+    Incidence::List incidencesWithAlarms(const QString &notebookUid,
+                                         const QString &uid) override;
+};
+//@endcond
+
+MultiCalendarStorage::MultiCalendarStorage(const QTimeZone &timezone)
+    : d(new Private(this, timezone))
+{
+}
+
+MultiCalendarStorage::~MultiCalendarStorage()
+{
+    delete d;
+}
+
+bool MultiCalendarStorage::setTimeZone(const QTimeZone &timezone)
+{
+    bool changed = false;
+    d->mTimezone = timezone;
+    for (QHash<QString, CalendarHandler>::Iterator it = d->mCalendars.begin();
+         it != d->mCalendars.end(); it++) {
+        changed = changed || (it->calendar()->timeZone() != timezone);
+        it->calendar()->setTimeZone(timezone);
+    }
+    return changed;
+}
+
+bool MultiCalendarStorage::open()
+{
+    for (const Notebook::Ptr &notebook : loadedNotebooks(&d->mDefaultNotebookUid)) {
+        QHash<QString, CalendarHandler>::Iterator it = d->mCalendars.find(notebook->uid());
+        if (it == d->mCalendars.end()) {
+            it = d->mCalendars.insert(notebook->uid(), CalendarHandler(d->mTimezone));
+        }
+        it->setNotebook(notebook);
+    }
+    return true;
+}
+
+bool MultiCalendarStorage::close()
+{
+    d->mCalendars.clear();
+    return true;
+}
+
+QString MultiCalendarStorage::instanceIdentifier(const QString &notebookUid,
+                                                 const Incidence &incidence)
+{
+    return instanceIdentifier(notebookUid, incidence.instanceIdentifier());
+}
+
+QString MultiCalendarStorage::instanceIdentifier(const QString &notebookUid,
+                                                 const QString &identifier)
+{
+    return QString::fromLatin1("%1::NBUID::%2").arg(notebookUid).arg(identifier);
+}
+
+static QPair<QString, QString> deserialiseIdentifier(const QString &instanceIdentifier)
+{
+    int at = instanceIdentifier.indexOf(QString::fromLatin1("::NBUID::"));
+    return at < 0 ? QPair<QString, QString>()
+                  : QPair<QString, QString>(instanceIdentifier.left(at), instanceIdentifier.right(at + 9));
+}
+
+bool MultiCalendarStorage::loadIncidenceInstance(const QString &instanceIdentifier)
+{
+    QPair<QString, QString> ids = deserialiseIdentifier(instanceIdentifier);
+    const QString notebookUid = ids.first;
+    QString uid = ids.second;
+    if (notebookUid.isEmpty() || uid.isEmpty()) {
+        qCWarning(lcMkcal) << "wrong instance identifier" << instanceIdentifier;
```

Comment:
Or maybe rather "invalid"? Wrong sounds like unexpected identifier.

Reply:
Ok, changed for "invalid". It looks better, indeed.

Reply:
The old text is btw having some other instances later on.

Reply:
You're right. I've changed the warning label in the three other cases.